### DISCUSSION TITLE
ILI9341 4-wire SPI dont work

### DISCUSSION THREAD

**Discussion by roesl** [arendst/Tasmota#17605] [Q&A] [answered]
I bought a ILI9341 4-wire SPI but It dont work.
![image](https://user-images.githubusercontent.com/23452986/210751832-7e576ca8-7ab3-423c-84ca-fd9f2a8ea58b.png)

console at start is:
00:00:00.002 HDW: ESP32-D0WDQ6 
00:00:00.028 UFS: FlashFS mounted with 304 kB free
00:00:00.057 CFG: Loaded from File, Count 17
00:00:00.066 QPC: Count 1
00:00:00.070 SPI: Hardware using GPIO14(CLK), GPIO13(MOSI) and GPIO12(MISO)
00:00:00.153 BRY: Berry initialized, RAM used=3627 bytes

![image](https://user-images.githubusercontent.com/23452986/210752355-c2ae7b4c-70f2-4685-949c-0c2ec9ac9fcb.png)

the display seems to be not recognised

Whats the problem ?
I use tasmota32-lvgl.bin

**Comment by sfromis** [comment]
You need an unused gpio configured as Option A3 too, to activate the universal display driver. And you need a `display.ini` with the setup for your display, like one from here https://github.com/arendst/Tasmota/tree/development/tasmota/displaydesc

More generally, Tasmota cannot detect SPI devices, you need to configure Tasmota to know what your device is.

**Comment by roesl** [comment]
no change 

![image](https://user-images.githubusercontent.com/23452986/210765603-cd005491-f45c-45b2-9859-12c9c158d107.png)
00:00:00.002 HDW: ESP32-D0WDQ6 
00:00:00.045 UFS: FlashFS mounted with 76 kB free
00:00:00.077 CFG: Loaded from File, Count 22
00:00:00.086 QPC: Count 1
00:00:00.090 SPI: Hardware using GPIO14(CLK), GPIO13(MOSI) and GPIO12(MISO)
00:00:00.171 BRY: Berry initialized, RAM used=3627 bytes
00:00:00.193 Project tasmota - Tasmota Version 12.3.1(display)-2_0_5_3(2022-12-16T10:31:14)

display.ini was loaded to file system.
Is the size of 225 kB for the display.ini correct ?

**Reply by sfromis** [reply]
> Is the size of 225 kB for the display.ini correct ?

By far not. 1000 times too big. Right-click on the "Raw" link (after clicking on the file), to download.

You should see `DSP: File descriptor used` and a bit later `DSP: Configured display 'ILI9341'` during boot (probably needs `Weblog 3` activated). Remember to power on the display (using front page toggle or power command) to start doing something. The Berry example from https://tasmota.github.io/docs/LVGL/#using-lvgl is usually a nice simple test to see it in action.

**Comment by Jason2866** [comment]
Not it is not correct. Do not use save as from your web browser. Click on file an use GitHub download button!

**Comment by roesl** [comment]
do I have to add the configuration to the first line of the display.ini ?

**Reply by sfromis** [reply]
No, most of the time it should be fine as is. And be very careful if you use Windows, many text editing tools will make the file unusable. Recommended is to download the file via the "Raw" link, as I posted, and upload it to the ESP32, without using copy/paste anytime.

**Reply by roesl** [reply]
the text editor was one problem

**Comment by roesl** [comment]
now I get 
00:00:00.003 HDW: ESP32-D0WDQ6 
00:00:00.047 UFS: FlashFS mounted with 304 kB free
00:00:00.083 CFG: Loaded from File, Count 18
00:00:00.092 QPC: Count 1
00:00:00.096 SPI: Hardware using GPIO17(CLK), GPIO22(MOSI) and GPIO21(MISO)
00:00:00.229 BRY: Berry initialized, RAM used=5008 bytes
00:00:00.692 RSL: RESULT = {"POWER1":"ON"}
00:00:00.693 RSL: POWER1 = ON
00:00:00.701 LVG: LVGL initialized
00:00:00.805 DSP: Configured display 'ILI9341'
00:00:00.819 Project tasmota - Tasmota Version 12.3.1(lvgl-haspmota)-2_0_5_3(2022-12-16T10:31:03)

and
13:28:04.544 RSL: RESULT = {"Display":{"Model":17,"Type":0,"Width":240,"Height":320,"Mode":0,"Dimmer":10,"Size":1,"Font":1,"Rotate":0,"Invert":0,"Refresh":2,"Cols":[16,8],"Rows":2}}

but nothing appears at the display

**Reply by Jason2866** [reply]
Increase brightness by doing `Displaydimmer 100`
Restart device, if working you see for a short time a small Tasmota logo.

**Reply by sfromis** [reply]
Right, that initial `"Dimmer":10` may not be very visible....

Your messages does not indicate that you followed my advice about powering on the display. Since you have a relay configured too, that would be Power1, and thus I'd expect the display to become Power2. Or use the front page toggle.

I'd suggest to also use the Backlight config and the corresponding display pin `LED`, but if you do not, you should also wire the display for backlight on. I think that the Tasmota power on is still needed, but not 100% sure.

**Reply by gemu2015** [reply]
Did you forget some pins?
you also need spi_dc and backlight and reset

**Comment by roesl** [comment, marked as answer]
its working now :)

Thanks to all !!

**Comment by roesl** [comment]
the problem was the display.ini which i copied to a windows editor and uploaded it to filesystem.
since this didnt work, I downloaded the display.ini from the browser and got a file of 225 kB :(

I didnt find a download button for the file itself .
But copy the raw file and create a new file in ESP filesystem was the solution.

**Reply by sfromis** [reply]
The reason for my specifically and repeatedly advicing to right-click the "Raw" button was that the "Download" button is only for binary files, not for text files. But with Windows nearby, text files for data usage needs to be treated as if it was binary files, to stop Windows from screwing it up.
